Content.Shared/Inventory/InventorySystem.Slots.cs

```diff
@@ -0,0 +1,116 @@
+
+namespace Content.Shared.Inventory;
+
+public partial class InventorySystem : EntitySystem
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+    public bool TryGetSlotContainer(EntityUid uid, string slot, [NotNullWhen(true)] out ContainerSlot? containerSlot, [NotNullWhen(true)] out SlotDefinition? slotDefinition,
+        InventoryComponent? inventory = null, ContainerManagerComponent? containerComp = null)
+    {
+        containerSlot = null;
+        slotDefinition = null;
+        if (!Resolve(uid, ref inventory, ref containerComp, false))
+            return false;
+
+        if (!TryGetSlot(uid, slot, out slotDefinition, inventory: inventory))
+            return false;
+
+        if (!containerComp.TryGetContainer(slotDefinition.Name, out var container))
+        {
+            containerSlot = containerComp.MakeContainer<ContainerSlot>(slotDefinition.Name);
+            return true;
+        }
+
+        if (container is not ContainerSlot containerSlotChecked) return false;
+
+        containerSlot = containerSlotChecked;
+        return true;
+    }
+
+    public bool HasSlot(EntityUid uid, string slot, InventoryComponent? component = null) =>
+        TryGetSlot(uid, slot, out _, component);
+
+    public bool TryGetSlot(EntityUid uid, string slot, [NotNullWhen(true)] out SlotDefinition? slotDefinition, InventoryComponent? inventory = null)
+    {
+        slotDefinition = null;
+        if (!Resolve(uid, ref inventory, false))
+            return false;
+
+        if (!_prototypeManager.TryIndex<InventoryTemplatePrototype>(inventory.TemplateId, out var templatePrototype))
+            return false;
+
+        slotDefinition = templatePrototype.Slots.FirstOrDefault(x => x.Name == slot);
+        return slotDefinition != default;
+    }
+
+    public bool TryGetContainerSlotEnumerator(EntityUid uid, out ContainerSlotEnumerator containerSlotEnumerator, InventoryComponent? component = null)
+    {
+        containerSlotEnumerator = default;
+        if (!Resolve(uid, ref component, false))
+            return false;
+
+        containerSlotEnumerator = new ContainerSlotEnumerator(uid, component.TemplateId, _prototypeManager, this);
+        return true;
+    }
+
+    public bool TryGetSlots(EntityUid uid, [NotNullWhen(true)] out SlotDefinition[]? slotDefinitions, InventoryComponent? inventoryComponent = null)
+    {
+        slotDefinitions = null;
+        if (!Resolve(uid, ref inventoryComponent, false))
+            return false;
+
+        if (!_prototypeManager.TryIndex<InventoryTemplatePrototype>(inventoryComponent.TemplateId, out var templatePrototype))
+            return false;
+
+        slotDefinitions = templatePrototype.Slots;
+        return true;
+    }
+
+    public SlotDefinition[] GetSlots(EntityUid uid, InventoryComponent? inventoryComponent = null)
+    {
+        if (!Resolve(uid, ref inventoryComponent, false)) throw new InvalidOperationException();
```

Comment:
If it's just gonna throw why not just make it a required argument?
The function is only called in one location anyway and that passes it in.

Reply:
good point tbh